wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivityManager.java

```diff
@@ -165,14 +191,282 @@ protected long getIdleTimeout(String wsId) {
+            runningWsId);
+        try {
+          createMissingActivityRecord(runningWsId);
+        } catch (NotFoundException e) {
+          LOG.error(
+              "Detected a running workspace {} but could not find" + " its record.",
+              runningWsId,
+              e);
+        } catch (ConflictException e) {
+          LOG.debug(
+              "Activity record created while we were trying to rectify its absence for a running"
+                  + " workspace {}.",
+              runningWsId);
+        }
+      } else {
+        // what are we trying to handle here?
+        // 1) make sure that the activity record has a valid created time
+        // 2) make sure activity record as at least 1 activity record - if it lacks last_running,
+        //    we look at last_starting and if even that is not there we make sure there is at least
+        //    created_time - the last_* are not mandatory, but the created_time should be present
+        // 3) absence of last_running - this can either happen very shortly after the workspace
+        //    has started and the schedule of this method coincided with it starting. or there has
+        //    been some kind of problem.
+        // 4) absence of expiry - again this should only happen shortly after the workspace has
+        //    started if the schedule of this method managed to read the activity before the expiry
+        //    time has been persisted in the event handler. Otherwise it is a problem.
+
+        if (activity.getCreated() == null) {
+          try {
+            rectifyCreatedTime(activity);
+          } catch (NotFoundException e) {
+            LOG.error(
+                "Detected a running workspace {} but could not find" + " its record.",
+                runningWsId,
+                e);
+          }
+        }
+
+        // let's use a single value for the current time in all the code below
+        long now = clock.millis();
+
+        boolean noLastRunningTime = activity.getLastRunning() == null;
+
+        // this value is the last recorded activity of any kind on the workspace.
+        // Even though we tried to recover the created_time in the code above, it might still happen
+        // that we failed to do that and that no other activity exists on the workspace.
+        // That's why in the code below we still have to account for the possibility of this value
+        // being null.
+        Long lastKnownActivity =
+            maxOf(
+                activity.getCreated(),
+                activity.getLastStarting(),
+                activity.getLastRunning(),
+                activity.getLastStopping(),
+                activity.getLastStopped());
+
+        if (noLastRunningTime) {
+          rectifyNoLastRunningTime(runningWsId, activity, now, lastKnownActivity);
+        } else if (lastKnownActivity != null && lastKnownActivity > activity.getLastRunning()) {
```

Comment:
This is a wrong check.

Reply:
Actually, no, it is correct :smile: We're talking about a workspace that is running - i.e. it is neither starting, stopping, nor stopped according to what the runtime thinks.